pow/src/main/java/tech/pegasys/artemis/pow/DepositRequestManager.java

```diff
@@ -0,0 +1,222 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import tech.pegasys.artemis.pow.api.DepositEventChannel;
+import tech.pegasys.artemis.pow.contract.DepositContract;
+import tech.pegasys.artemis.pow.event.Deposit;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class DepositRequestManager {
+
+  private final Eth1Provider eth1Provider;
+  private final DepositEventChannel depositEventChannel;
+  private final DepositContract depositContract;
+  private static final Logger LOG = LogManager.getLogger();
+  private final AsyncRunner asyncRunner;
+
+  private volatile Disposable newBlockSubscription;
+  private boolean active = false;
+  private boolean requestQueued = false;
+  private BigInteger latestCanonicalBlockNumber;
+  private BigInteger latestSuccessfullyQueriedBlockNumber = BigInteger.ZERO;
+
+  public DepositRequestManager(
+      Eth1Provider eth1Provider,
+      AsyncRunner asyncRunner,
+      DepositEventChannel depositEventChannel,
+      DepositContract depositContract) {
+    this.eth1Provider = eth1Provider;
+    this.asyncRunner = asyncRunner;
+    this.depositEventChannel = depositEventChannel;
+    this.depositContract = depositContract;
+  }
+
+  public void start() {
+    newBlockSubscription =
+        eth1Provider
+            .getLatestBlockFlowable()
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(this::updateLatestCanonicalBlockNumber, this::onSubscriptionFailed);
+  }
+
+  private void onSubscriptionFailed(Throwable err) {
+    Disposable subscription = newBlockSubscription;
+    if (subscription != null) {
+      subscription.dispose();
+    }
+    LOG.warn("New block subscription failed, retrying.", err);
+    start();
+  }
+
+  public void stop() {
+    newBlockSubscription.dispose();
+  }
+
+  private synchronized void updateLatestCanonicalBlockNumber(
+      BigInteger latestCanonicalBlockNumber) {
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    getLatestDeposits();
+  }
+
+  private void getLatestDeposits() {
+    final BigInteger latestCanonicalBlockNumberAtRequestStart;
+    final BigInteger latestSuccessfullyQueriedBlockNumberAtRequestStart;
+    synchronized (DepositRequestManager.this) {
+      if (active) {
+        requestQueued = true;
+        return;
+      } else {
+        active = true;
+        requestQueued = false;
+      }
```

Comment:
nit: Because we return in the true case this doesn't need to be inside an else block.

Reply:
Done.